components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java

```diff
@@ -580,6 +614,266 @@ protected void initiateAuthenticationRequest(HttpServletRequest request, HttpSer
+                                                          Map<String, String> authenticatorProperties, String scopes) {
+
+        // Get the application requested scopes for the federated tokens.
+        String requestedScopesForTokenSharing = getRequestedScopesForTokenSharing(context);
+
+        // Validating the application requested scopes by the authenticator allowed scopes for federated token sharing.
+        Set<String> validScopesForTokenSharing = validateScopeForTokenSharing(
+                authenticatorProperties.get(OIDCAuthenticatorConstants.FEDERATED_TOKEN_ALLOWED_SCOPE),
+                requestedScopesForTokenSharing);
+
+        if (CollectionUtils.isEmpty(validScopesForTokenSharing)) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("No matching scopes found for federated token sharing.");
+            }
+            return scopes;
+        }
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Valid scopes found for federated token sharing: " + validScopesForTokenSharing +
+                    ", IDP: " + getFederatedAuthenticatorName(context));
+        }
+        /*
+        Remove the duplicate scopes among the validated scopes for federated token sharing and the existing scopes
+        of the authenticators.
+         */
+        scopes = removeDuplicateScopes(scopes, validScopesForTokenSharing);
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("The scopes for the IDP: " + getFederatedAuthenticatorName(context) + " : " + scopes +
+                    " after considering federated token sharing.");
+        }
+        return scopes;
+    }
+
+    /**
+     * This method is used to remove the duplicate scopes.
+     *
+     * @param scopes                     The scopes defined in the authenticator. i.e. "openid email profile"
+     * @param validScopesForTokenSharing The validated scopes requested by the application and the allowed scopes
+     *                                   for the token sharing.
+     * @return The scopes after removing the duplicate scopes.
+     */
+    private String removeDuplicateScopes(String scopes, Set<String> validScopesForTokenSharing) {
+
+        if (StringUtils.isBlank(scopes)) {
+            scopes = StringUtils.join(validScopesForTokenSharing, SPACE);
+        }
+
+        Set<String> scopeSet = new HashSet<>(Arrays.asList(scopes.split(SPACE_REGEX)));
+        scopeSet.addAll(validScopesForTokenSharing);
+
+        scopes = StringUtils.join(scopeSet, SPACE);
+        return scopes;
+    }
+
+    /**
+     * This method returns the scopes requested by the application for the federated tokens.
+     *
+     * @param context The authentication context.
+     * @return The scopes requested by the application for token sharing.
+     */
+    private String getRequestedScopesForTokenSharing(AuthenticationContext context) {
+
+        String authenticatorName = getFederatedAuthenticatorName(context);
```

Comment:
since this is only used in debug logs lets call the method directly in the log

Reply:
Fixed d2b40f9255b4a5ef3b8f9d44bd58ccdba9f50f70